nre/bpf/nodefw/nodefw/src/fwmap.rs

```diff
@@ -1,54 +1,97 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use crate::time::get_ktime_get_ns;
-use anyhow::Error;
+use crate::server::BlockAddress;
+use crate::time::{get_ktime_get_ns, ttl};
+use anyhow::Result;
 use aya::maps::MapData;
 use aya::{maps::HashMap, Bpf};
 use log::{error, info};
 use nodefw_common::Rule;
-use std::cell::RefCell;
 use std::cmp::Ordering;
 use std::collections::BinaryHeap;
 use std::net::{IpAddr, Ipv6Addr};
 use std::str::FromStr;
+use std::sync::{Arc, RwLock};
 use std::time::Duration;
 use tokio_util::sync::CancellationToken;
 
 pub struct Firewall {
-    inner: HashMap<MapData, [u8; 16usize], Rule>,
-    expirations: BinaryHeap<TtlRecord>,
+    inner: Arc<RwLock<HashMap<MapData, [u8; 16usize], Rule>>>,
+    expirations: Arc<RwLock<BinaryHeap<TtlRecord>>>,
 }
 impl Firewall {
     pub fn new(map_name: &str, bpf: &mut Bpf) -> Self {
         Self {
-            inner: HashMap::try_from(bpf.take_map(map_name).unwrap()).unwrap(),
-            expirations: BinaryHeap::new(),
+            inner: Arc::new(RwLock::new(
+                HashMap::try_from(bpf.take_map(map_name).unwrap()).unwrap(),
+            )),
+            expirations: Arc::new(RwLock::new(BinaryHeap::new())),
         }
     }
-    pub fn add(&mut self, ip: &str, rule: Rule) -> Result<(), Error> {
-        let parsed_ip = match IpAddr::from_str(ip) {
-            Ok(IpAddr::V4(v)) => v.to_ipv6_compatible(),
-            Ok(IpAddr::V6(v)) => v,
-            Err(e) => {
-                error!("{}", e);
-                return Err(e.into());
+    pub fn block_addresses(&mut self, addresses: Vec<BlockAddress>) -> Result<()> {
```

Comment:
I think we should make this API more forward looking by supporting specifying whether each address in the request is for a connection IP or a proxy IP. Will discus offline the justification

Reply:
chatted, deferring this for now.